benchmark/src/benchmark/measure.clj

```diff
@@ -11,84 +12,106 @@
-    (concat query-times connection-times transaction-times)))
+(defn init-tx [entity-count conn]
+  (d/transact conn c/schema)
+  (if (pos? entity-count)
+    (doseq [entities (->> (repeatedly entity-count (partial c/rand-entity entity-count))
+                          (partition 100000))]
+      (d/transact conn (vec entities)))
+    []))
+
+(defn time-context-map
+  ([t config function entity-count datom-count]
+   (time-context-map t config function entity-count datom-count nil))
+  ([t config function entity-count datom-count exec]
+   (cond-> {:time t
+            :context {:dh-config config
+                      :function function
+                      :db-entities entity-count
+                      :db-datoms datom-count}}
+     (some? exec) (assoc-in [:context :execution] exec))))
+
+(defn measure-connection-time [iterations unique-cfg simple-cfg entity-count datom-count]
+  (log/info (str "Measuring connection time on database with config named '" (:name simple-cfg) ", "
+                 (count c/schema) " attributes in entity, and " entity-count " entities in database"))
+  (for [_ (range iterations)]
+    (let [timed-conn (timed (d/connect unique-cfg))]
+      (d/release (:res timed-conn))
+      (time-context-map (:t timed-conn) simple-cfg :connection entity-count datom-count))))
+
+(defn measure-query-times
+  [{:keys [iterations data-found-opts query data-types]} initial-tx conn config entity-count datom-count]
+  (log/info (str "Measuring query times on database with config named '" (:name config) ", "
+                 (count c/schema) " attributes in entity, and " entity-count " entities in database"))
+  (let [data-found (case data-found-opts
+                     true [true]
+                     false [false]
+                     :all [true false])
+        queries (if (pos? (count initial-tx))
+                  (c/all-queries @conn initial-tx data-types data-found)
+                  (c/non-var-queries @conn data-types (count initial-tx)))
+        filtered-queries (if (= query :all)
+                           queries
+                           (filter #(= query (:function %)) queries))]
+    (flatten
+     (for [{:keys [function query details]} filtered-queries]
+       (do
+         (log/debug (str " Querying with " function " using " details "..."))
+         (for [_ (range iterations)]
+           (time-context-map (:t (timed (d/q query @conn)))
+                             config function entity-count datom-count details)))))))
+
+(defn measure-transaction-times
+  [{:keys [iterations tx-entity-counts]} conn config entity-count datom-count]
+  (log/info (str "Measuring transaction times on database with config named '" (:name config) ", "
+                 (count c/schema) " attributes in entity, and " entity-count " entities in database"))
+  (for [_ (range iterations)    ;; sampling for given entity-count
+        tx-entities tx-entity-counts
+        _ (range iterations)]   ;; sampling for each tx-entities
```

Comment:
The first corresponds to sampling each config, while the second corresponds to number of entities in each transaction (when benchmarking transactions). In principle, the two numbers should be independent, but they coincided in the benchmarking code and I stuck with the status quo.

Reply:
I see where you are coming from, but in the original version the database changed between the iterations, so it made at least a little sense to have both repetitions there because there was actual sampling going on (although even then one admittedly could have been skipped). In fact, this shows that the first iteration-loop should go back to its original spot, because otherwise the benchmarking is done on a single database. The reason why it makes sense to test different databases at the moment are mainly the caches on different levels that will skew the benchmark results especially for queries. As I understand we will be working on making the ones we have an influence on configurable, so we will be able to bypass this problem. For now we should stick to the momentary solution even though it is slower.  What we can do though is to actually make those two iteration counts different parameters, so at least one parameter to configure the amount of times a database with the same number of entities is built and tested, and one parameter to configure the amount of times a specific function is tested on a database. This way we would still have the option to decide what is best for a specific benchmarking case.

Reply:
On second thought, perhaps [sampling only each DB and transaction entity count combination](https://github.com/replikativ/datahike/pull/496/commits/7ef47a55881f03d7e486980e97afb0b47bbc2136) makes more sense? I changed the code to do that. Do you think the database should still be reinstantiated for each sample? Or that we should still have two separate parameters for DB and transaction entity count?

Reply:
Leave the code on line 73-74 as you have it and the second loop has to got to its original spot, i.e. line 137 now (see comment below), so there will be an actual sampling going on. With that parameter the user will have the choice to either run the benchmarks quickly, by setting _iterations_ to the number of samples wanted and _db-samples_ to 1, which tests a function on the same database _iterations_ times accepting that queries might seem quicker because caches are being used, or the user is interested in query times without caching and so should set _db-samples_ to the number of samples wanted and _iterations_ to 1 which will result in the same number of samples, but it will run slower because the database is built up again for each sample, but it should yield more stable results because it circumvents the caching issue.